binary_transparency/firmware/internal/client/follower.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/binary_transparency/firmware/api"
+	"github.com/google/trillian-examples/binary_transparency/firmware/internal/crypto"
+	"github.com/google/trillian-examples/binary_transparency/firmware/internal/verify"
+	"github.com/google/trillian/merkle/logverifier"
+)
+
+type ErrConsistency struct {
+	Golden, Latest api.LogCheckpoint
+}
+
+func (e ErrConsistency) Error() string {
+	return fmt.Sprintf("failed to verify consistency proof from %s to %s", e.Golden, e.Latest)
+}
+
+type ErrInclusion struct {
+	Checkpoint api.LogCheckpoint
+	Proof      api.InclusionProof
+}
+
+func (e ErrInclusion) Error() string {
+	return fmt.Sprintf("failed to verify inclusion proof %s in root %s", e.Proof, e.Checkpoint)
+}
+
+// LogEntry wraps up a leaf value with its position in the log.
+type LogEntry struct {
+	Root  api.LogCheckpoint
+	Index uint64
+	Value api.SignedStatement
+}
+
+// LogFollower follows a log for new data becoming available.
+type LogFollower struct {
+	c            ReadonlyClient
+	lv           logverifier.LogVerifier
+	pollInterval time.Duration
+	golden       api.LogCheckpoint
+}
+
+func NewLogFollower(c ReadonlyClient, pollInterval time.Duration, golden api.LogCheckpoint) LogFollower {
+	return LogFollower{
+		c:            c,
+		lv:           verify.NewLogVerifier(),
+		pollInterval: pollInterval,
+		golden:       golden,
+	}
+}
+
+// Checkpoints polls the log according to the configured interval, returning roots consistent
+// with the current golden checkpoint. Should any valid roots be found which are inconsistent
+// then an error is returned. The log being unavailable will just cause a retry, giving it
+// the benefit of the doubt.
+func (f *LogFollower) Checkpoints(ctx context.Context) (<-chan api.LogCheckpoint, <-chan error) {
+	ticker := time.NewTicker(f.pollInterval)
+
+	outc := make(chan api.LogCheckpoint, 1)
+	errc := make(chan error, 1)
+
+	// Push the starting checkpoint into the channel here to kick off the pipeline
+	outc <- f.golden
+
+	go func() {
```

Comment:
Just a question, mostly for my benefit.
If @line 85, if this is run as a separate thread, via go func, how the calling code inside Monitor or Witness wait or block for a poll ??

Reply:
`outc` and `errc` are buffered with a size of 1, which will prevent immediate blocking if only writing a single element. If we didn't buffer `outc` then this method would never get past L85 as currently written.

Calling code blocks on the channel returned by this method until elements are available. While the channel is empty, the code will block and wait for it. This is effectively the same pattern as before, except then channel was `ticker.C` instead of a custom channel with Checkpoints on it.

Reply:
Sure thanks, just wondering as this breaks the logic into two for loops (one in monitor and one in follower), can there be a race condition where the monitor thread is not complete and a new checkpoint is available on the channel, will the monitor context of checkpoint over-written in this use case?

Reply:
Maybe take a look at https://blog.golang.org/pipelines if you'd like to see more about Go concurrency, especially this "pipeline" style of processing.

Reply:
Sure will do!